staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache_interval_test.go

```diff
@@ -0,0 +1,428 @@
+		{endIndex: bufferSize, expected: true},
+		{endIndex: bufferSize + 1, expected: true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{endIndex: c.endIndex}
+		actual := wcib.isFull()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferIsEmpty(t *testing.T) {
+	cases := []struct {
+		startIndex int
+		endIndex   int
+		expected   bool
+	}{
+		{startIndex: 0, endIndex: 10, expected: false},
+		{startIndex: 5, endIndex: 20, expected: false},
+		{startIndex: 50, endIndex: 50, expected: true},
+	}
+
+	for _, c := range cases {
+		wcib := &watchCacheIntervalBuffer{
+			startIndex: c.startIndex,
+			endIndex:   c.endIndex,
+		}
+		actual := wcib.isEmpty()
+		if actual != c.expected {
+			t.Errorf("expected %v, got %v", c.expected, actual)
+		}
+	}
+}
+
+func TestIntervalBufferNext(t *testing.T) {
+	cases := []struct {
+		name   string
+		events []*watchCacheEvent
+	}{
+		{
+			name: "buffer has elements",
+			events: []*watchCacheEvent{
+				{Type: watch.Added, Object: makeTestPod("pod1", 1)},
+				{Type: watch.Added, Object: makeTestPod("pod2", 2)},
+				{Type: watch.Modified, Object: makeTestPod("pod3", 3)},
+			},
+		},
+		{
+			name:   "buffer is empty",
+			events: []*watchCacheEvent{},
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			wcib := bufferFromEvents(c.events)
+			for i := 0; i < len(c.events); i++ {
+				event, ok := wcib.next()
+				if err := verifyEvent(ok, event, c.events[i]); err != nil {
+					t.Error(err)
+				}
+			}
+			event, ok := wcib.next()
+			if err := verifyNoEvent(ok, event); err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+func TestFillBuffer(t *testing.T) {
+	cases := []struct {
+		name            string
+		numEventsToFill int
+	}{
+		{
+			name:            "no events to put in buffer",
+			numEventsToFill: 0,
+		},
+		{
+			name:            "less than bufferSize events to put in buffer",
+			numEventsToFill: 5,
+		},
+		{
+			name:            "equal to bufferSize events to put in buffer",
+			numEventsToFill: bufferSize,
+		},
+		{
+			name:            "greater than bufferSize events to put in buffer",
+			numEventsToFill: bufferSize + 5,
+		},
+	}
+
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			events := generateEvents(0, c.numEventsToFill)
+			wci := intervalFromEvents(events)
+
+			for i := 0; i < len(events); i++ {
+				if i%bufferSize == 0 {
+					wci.fillBuffer()
+				}
+				event, ok := wci.buffer.next()
+				if err := verifyEvent(ok, event, events[i]); err != nil {
+					t.Error(err)
+				}
+				// If we have already received bufferSize number of events,
+				// buffer should be empty and we should receive no event.
+				if i%bufferSize == bufferSize-1 {
+					event, ok := wci.buffer.next()
+					if err := verifyNoEvent(ok, event); err != nil {
+						t.Error(err)
+					}
+				}
+			}
+			// Buffer should be empty now, an additional fillBuffer()
```

Comment:
add verifyNoEvent before it

[in the for loop above you only change verifyNoEvent at the end if len(events)%bufferSize==bufferSize-1. We want to check that for other cases too.]